server/src/main/java/org/elasticsearch/action/bulk/BulkProcessor2.java

```diff
@@ -0,0 +1,427 @@
+         */
+        public Builder setMaxNumberOfRetries(int maxNumberOfRetries) {
+            assert maxNumberOfRetries >= 0;
+            this.maxNumberOfRetries = maxNumberOfRetries;
+            return this;
+        }
+
+        /**
+         * Sets the maximum number of bytes allowed in in-flight requests (both the BulkRequest being built up by the BulkProcessor and
+         * any BulkRequests sent to Retry2 that have not yet completed) before subsequent calls to add()result in
+         * EsRejectedExecutionException. Defaults to 50mb.
+         */
+        public Builder setMaxBytesInFlight(ByteSizeValue maxBytesInFlight) {
+            this.maxBytesInFlight = maxBytesInFlight;
+            return this;
+        }
+
+        /**
+         * Builds a new bulk processor.
+         */
+        public BulkProcessor2 build() {
+            return new BulkProcessor2(
+                consumer,
+                maxNumberOfRetries,
+                listener,
+                maxRequestsInBulk,
+                maxBulkSizeInBytes,
+                maxBytesInFlight,
+                flushInterval,
+                threadPool
+            );
+        }
+    }
+
+    /**
+     * @param consumer The consumer that is called to fulfil bulk operations. This consumer _must_ operate either very fast or
+     *                 asynchronously.
+     * @param listener The BulkProcessor2 listener that gets called on bulk events
+     * @param threadPool The threadpool used to schedule the flush task for this bulk processor, if flushInterval is not null.
+     * @return the builder for BulkProcessor2
+     */
+    public static Builder builder(
+        BiConsumer<BulkRequest, ActionListener<BulkResponse>> consumer,
+        Listener listener,
+        ThreadPool threadPool
+    ) {
+        Objects.requireNonNull(consumer, "consumer");
+        Objects.requireNonNull(listener, "listener");
+        return new Builder(consumer, listener, threadPool);
+    }
+
+    private final int maxActionsPerBulkRequest;
+    private final long maxBulkSizeBytes;
+    private final ByteSizeValue maxBytesInFlight;
+    /*
+     * This is the approximate total number of bytes in in-flight requests, both in the BulkRequest that it is building up and in all of
+     * the BulkRequests that it has sent to Retry2 that have not completed yet. If this number would exceeds maxBytesInFlight, then calls
+     * to add() will throw EsRejectedExecutionExceptions.
+     */
+    private final AtomicLong totalBytesInFlight = new AtomicLong(0);
+
+    /**
+     * This is a task (which might be null) that is scheduled at some pont in the future to flush the bulk request and start a new bulk
+     * request. This variable is read and written to from multiple threads, and is protected by mutex.
+     */
+    private volatile Scheduler.Cancellable cancellableFlushTask = null;
+
+    private final AtomicLong executionIdGen = new AtomicLong();
+
+    private final Logger logger;
```

Comment:
Echoing a nit from a zoom -- probably best to make sure all the loggers in this PR are static rather than instance.